src/main/scala/io/iohk/ethereum/merklepatriciatrie/MerklePatriciaTrie.scala

```diff
@@ -0,0 +1,487 @@
+package io.iohk.ethereum.merklepatriciatrie
+
+import io.iohk.ethereum.merklepatriciatrie.MerklePatriciaTrie.HashFn
+import io.iohk.ethereum.rlp.RLPImplicits._
+import io.iohk.ethereum.rlp.{decode => decodeRLP, encode => encodeRLP, _}
+
+import scala.annotation.tailrec
+
+object MerklePatriciaTrie {
+
+  case class MPTException(message: String) extends RuntimeException(message)
+
+  type HashFn = Array[Byte] => Array[Byte]
+
+  private val PairSize: Byte = 2
+  private[merklepatriciatrie] val ListSize: Byte = 17
+
+  def apply[K, V](source: DataSource, hashFn: HashFn)
+                 (implicit kSerializer: ByteArraySerializable[K], vSerializer: ByteArraySerializable[V])
+  : MerklePatriciaTrie[K, V] = new MerklePatriciaTrie[K, V](None, source, hashFn)(kSerializer, vSerializer)
+
+  def apply[K, V](rootHash: Array[Byte], source: DataSource, hashFn: HashFn)
+                 (implicit kSerializer: ByteArraySerializable[K], vSerializer: ByteArraySerializable[V])
+  : MerklePatriciaTrie[K, V] = new MerklePatriciaTrie[K, V](Some(rootHash), source, hashFn)(kSerializer, vSerializer)
+
+  private def getNode(nodeId: Array[Byte], source: DataSource)(implicit nodeDec: RLPDecoder[Node]): Node = {
+    val nodeEncoded = tryGetNode(nodeId, source)
+    decodeRLP[Node](nodeEncoded)
+  }
+
+  private def tryGetNode(key: Array[Byte], source: DataSource): Array[Byte] =
```

Comment:
@adamsmo i think this is a good point and should be changed. It originally returned a try and we left it as it is.